src/monitor/group_state_machine.c

```diff
@@ -295,14 +341,533 @@ ProceedGroupState(AutoFailoverNode *activeNode)
+		{
+			char message[BUFSIZE];
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to fast_forward "
+				"after %s:%d converged to wait_cascade.",
+				candidateNode->nodeName, candidateNode->nodePort,
+				activeNode->nodeName, activeNode->nodePort);
+
+			AssignGoalState(candidateNode,
+							REPLICATION_STATE_FAST_FORWARD, message);
+
+			return true;
+		}
+
+		/*
+		 * The candidate had to fast forward with the activeNode, grabbing WAL
+		 * bits that only this node had. Now that's done. The activeNode has to
+		 * follow the new primary that's being promoted.
+		 */
+		if (IsCurrentState(activeNode, REPLICATION_STATE_WAIT_CASCADE) &&
+			IsCurrentState(candidateNode, REPLICATION_STATE_STOP_REPLICATION))
+		{
+			char message[BUFSIZE];
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to secondary "
+				"after %s:%d converged to stop_replication.",
+				activeNode->nodeName, activeNode->nodePort,
+				candidateNode->nodeName, candidateNode->nodePort);
+
+			AssignGoalState(activeNode, REPLICATION_STATE_SECONDARY, message);
+
+			return true;
+		}
+
+		/*
+		 * When the candidate is done fast forwarding the locally missing WAL
+		 * bits, it can be promoted.
+		 */
+		if (IsCurrentState(activeNode, REPLICATION_STATE_FAST_FORWARD))
+		{
+			char message[BUFSIZE];
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to prepare_promotion",
+				activeNode->nodeName, activeNode->nodePort);
+
+			AssignGoalState(activeNode,
+							REPLICATION_STATE_PREPARE_PROMOTION,
+							message);
+
+			return true;
+		}
+
+		/*
+		 * When the activeNode is "just" another standby, it's time to follow
+		 * the new primary as soon as our candidate reaches stop_replication.
+		 */
+		if (IsCurrentState(activeNode, REPLICATION_STATE_REPORT_LSN) &&
+			IsCurrentState(candidateNode, REPLICATION_STATE_STOP_REPLICATION))
+		{
+			char message[BUFSIZE];
+
+			LogAndNotifyMessage(
+				message, BUFSIZE,
+				"Setting goal state of %s:%d to secondary "
+				"after %s:%d converged to stop_replication.",
+				activeNode->nodeName, activeNode->nodePort,
+				candidateNode->nodeName, candidateNode->nodePort);
+
+			AssignGoalState(activeNode, REPLICATION_STATE_SECONDARY, message);
+
+			return true;
+		}
+
+		/* when we have a candidate, we don't go through finding a candidate */
+		return false;
+	}
+
+	/*
+	 * We reach this code when we don't have an healthy primary anymore, it's
+	 * been demoted or is draining now. Most probably it's dead. We have
+	 * collected the last received LSN from all the standby nodes in SECONDARY
+	 * state, and now we need to select one of the failover candidates.
+	 *
+	 * The selection is based on candidatePriority. If the candidate with the
+	 * higher priority doesn't have the most recent LSN, we have it fetch the
+	 * missing WAL bits from one of the standby which did receive them.
+	 */
+	if (reportedLSNCount == candidateCount)
```

Comment:
If I understand correctly, we require that all standbys be up to continue with a failover. What would be a way to improve that?

Reply:
All standby nodes that are currently considered as a potential candidate need to have reported their LSN. From one call to `node_active` to the next, that candidate list might change. I have improved how we build the list itself, and then the comments around this test. Oh and fixed it so that we prevent progress when candidateCount is zero...